From: Roman Mamedov<roman@rm.pp.ru>
Subject: workaround for hardware bug in SB600 chipset
References: bnc#419250

magic bitbanging to solve hardware issue with more than one attached
device

Signed-off-by: Oliver Neukum <oneukum@suse.de>

--- a/drivers/usb/host/ehci-pci.c	2008-11-21 22:23:07.000000000 +0500
+++ b/drivers/usb/host/ehci-pci.c	2008-11-22 02:44:53.735604001 +0500
@@ -169,17 +169,18 @@
 		}
 		break;
 	case PCI_VENDOR_ID_ATI:
-		/* SB700 old version has a bug in EHCI controller,
+		/* SB600 and an old version of SB700 have a bug in EHCI controller,
 		 * which causes usb devices lose response in some cases.
 		 */
-		if (pdev->device == 0x4396) {
+		if ((pdev->device == 0x4386) || (pdev->device == 0x4396)) {
 			p_smbus = pci_get_device(PCI_VENDOR_ID_ATI,
 						 PCI_DEVICE_ID_ATI_SBX00_SMBUS,
 						 NULL);
 			if (!p_smbus)
 				break;
 			rev = p_smbus->revision;
-			if ((rev == 0x3a) || (rev == 0x3b)) {
+			if ((pdev->device == 0x4386) || (rev == 0x3a) || (rev == 0x3b)) {
+                                ehci_info(ehci, "applying AMD SB600/SB700 USB freeze workaround\n");
 				u8 tmp;
 				pci_read_config_byte(pdev, 0x53, &tmp);
 				pci_write_config_byte(pdev, 0x53, tmp | (1<<3));
